qupath-core/src/main/java/qupath/lib/images/servers/ColorTransforms.java

```diff
@@ -230,182 +253,205 @@ public String toString() {
+
+		/**
+		 * Get the channel number to extract (0-based index).
+		 */
+		public int getChannelNumber() {
+			return channel;
 		}
-		
-		
-		
 	}
-	
-	
-	static class ExtractChannel implements ColorTransform {
-		
-		private int channel;
-		
-		ExtractChannel(int channel) {
-			this.channel = channel;
+
+	static class ExtractChannelByName implements ColorTransform {
+
+		private final String channelName;
+
+		public ExtractChannelByName(String channel) {
+			this.channelName = channel;
 		}
-	
+
 		@Override
 		public float[] extractChannel(ImageServer<BufferedImage> server, BufferedImage img, float[] pixels) {
 			pixels = ensureArrayLength(img, pixels);
-			return img.getRaster().getSamples(0, 0, img.getWidth(), img.getHeight(), channel, pixels);
+			int c = getChannelNumber(server);
+			if (c >= 0) {
+				return img.getRaster().getSamples(0, 0, img.getWidth(), img.getHeight(), c, pixels);
+			}
+			throw new IllegalArgumentException("No channel found with name " + channelName);
 		}
-		
+
 		@Override
 		public String getName() {
-			return "Channel " + (channel + 1);
+			return channelName;
 		}
-		
+
 		@Override
-		public String toString() {
-			return getName();
-		}
-		
-		/**
-		 * Get the channel number to extract (0-based index).
-		 * @return
-		 */
-		public int getChannelNumber() {
-			return channel;
+		public boolean supportsImage(ImageServer<BufferedImage> server) {
+			return getChannelNumber(server) >= 0;
 		}
 
 		@Override
-		public boolean supportsImage(ImageServer<BufferedImage> server) {
-			return channel < server.nChannels();
+		public String toString() {
+			return getName();
 		}
 
 		@Override
 		public int hashCode() {
 			final int prime = 31;
 			int result = 1;
-			result = prime * result + channel;
+			result = prime * result + ((channelName == null) ? 0 : channelName.hashCode());
 			return result;
 		}
 
 		@Override
 		public boolean equals(Object obj) {
 			if (this == obj)
 				return true;
-			if (obj == null)
-				return false;
-			if (getClass() != obj.getClass())
-				return false;
-			ExtractChannel other = (ExtractChannel) obj;
-			if (channel != other.channel)
+			if (!(obj instanceof ExtractChannelByName extractChannel))
 				return false;
-			return true;
+			return Objects.equals(channelName, extractChannel.channelName);
+		}
+
+		/**
+		 * Get the channel name to extract. Can be null
+		 */
+		public String getChannelName() {
+			return channelName;
+		}
+
+		private int getChannelNumber(ImageServer<BufferedImage> server) {
+			return server.getMetadata().getChannels()
+					.stream()
+					.map(ImageChannel::getName)
+					.toList()
+					.indexOf(channelName);
 		}
-		
-		
-		
 	}
-	
-	static class ExtractChannelByName implements ColorTransform {
-		
-		private String channelName;
-		
-		ExtractChannelByName(String channel) {
-			this.channelName = channel;
+
+	static class LinearCombinationChannel implements ColorTransform {
+
+		private final Map<String, Float> channelNamesToCoefficients;
```

Comment:
I used `float` because `extractChannel` returns an array of `floats`. Do you think it's better to have `double` for intermediary operations still?

Reply:
Yes, I think so. Using `float` for the output halves the memory requirements compared to `double`, but doing intermediate calculations with `double` should improve accuracy with no lasting cost in terms of memory.

For the map, `Map<String, Number>` makes sense as well since we can always use `Number.doubleValue()` when required - which would even permit `BigDecimal`. Although any accuracy of `BigDecimal` would be lost during the calculation, so I don't think that is necessary.